### PR TITLE
Migrate `snaps-simulator` to `snaps-monorepo`

### PR DESCRIPTION
**Pull request by Mrtenz**
This migrates `snaps-simulator` from a separate repo to `snaps-monorepo`.

The source is almost the same, with some minor changes:

- The `tsconfig.json` was changed to match this repository.
- The `SubjectMetadataController` was added to allow snaps to install with the latest `SnapController`.
- The `processSnapPermissions` function was removed, and is now imported from `snaps-controllers`.

### CODE REVIEW

File: packages/snaps-simulator/src/features/simulation/slice.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+      state,
+      action: PayloadAction<GenericPermissionController>,
+    ) {
+      state.permissionController = action.payload as any;
+    },
+    setSubjectMetadataController(
+      state,
+      action: PayloadAction<SubjectMetadataController>,
+    ) {
+      state.subjectMetadataController = action.payload as any;
```
Comment (FrederikBolding) [inline packages/snaps-simulator/src/features/simulation/slice.ts:75]: Does this need a cast?

File: packages/snaps-simulator/src/features/simulation/snap-permissions.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import {
+  restrictedMethodPermissionBuilders,
+  selectHooks,
+} from '@metamask/rpc-methods';
+import { endowmentPermissionBuilders } from '@metamask/snaps-controllers';
+import { DEFAULT_ENDOWMENTS } from '@metamask/snaps-utils';
+
+export const ExcludedSnapEndowments = Object.freeze(['endowment:keyring']);
+
+export const buildSnapEndowmentSpecifications = () =>
```
Comment (FrederikBolding) [inline packages/snaps-simulator/src/features/simulation/snap-permissions.ts:11]: Should be exported from `snaps-controllers` - but we can do that in a separate PR. Create an issue perhaps?

File: packages/snaps-simulator/.eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+          {
+            allow: ['Text'],
+          },
+        ],
+
+        'react/display-name': 'off',
+        'react/prop-types': 'off',
+
+        // TODO: Investigate why this is needed.
+        'node/no-unpublished-import': 'off',
```
Comment (FrederikBolding) [inline packages/snaps-simulator/.eslintrc.js:34]: Is this only a problem in the Webpack config? If so, maybe we just disable for that one file?

File: packages/snaps-simulator/webpack.config.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      },
+      {
+        test: /\.css$/u,
+        use: ['style-loader', 'css-loader'],
+      },
+    ],
+  },
+  /* eslint-disable @typescript-eslint/naming-convention */
+  externals: {
+    'node:module': 'commonjs module',
```
Comment (FrederikBolding) [inline packages/snaps-simulator/webpack.config.ts:45]: I wonder if we can configure it to not bring in all these unused deps

File: packages/snaps-simulator/package.json [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    "jest-environment-jsdom": "^29.5.0",
+    "jest-fetch-mock": "^3.0.3",
+    "jest-it-up": "^2.0.0",
+    "lint-staged": "^12.4.1",
+    "monaco-editor-webpack-plugin": "^7.0.1",
+    "prettier": "^2.7.1",
+    "prettier-plugin-packagejson": "^2.2.11",
+    "process": "^0.11.10",
+    "react-refresh": "^0.14.0",
+    "readable-stream": "^4.4.0",
```
Comment (FrederikBolding) [inline packages/snaps-simulator/package.json:113]: Realizing now that we are bringing in a newer version of readable-stream than Browserify. Do we want to match its versions for the things we polyfill? 

It will also potentially simplify our LM policy diffs.

File: package.json [hunk 6]
```diff
@@ -72,6 +73,7 @@
     "eslint-plugin-node": "^11.1.0",
     "eslint-plugin-prettier": "^4.2.1",
     "execa": "^5.1.1",
+    "favicons": "^7.1.2",
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply package.json:76]: Yes, since they require a script to be run. When installing in the package, `allow-scripts` won't run.

File: packages/snaps-controllers/jest.config.js [hunk 7]
```diff
@@ -9,5 +9,7 @@ module.exports = deepmerge(baseConfig, {
   testTimeout: 10000,
 
   // This is required for `jest-fetch-mock` to work.
+  automock: false,
   resetMocks: false,
+  restoreMocks: false,
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply packages/snaps-controllers/jest.config.js:14]: Required for `jest-fetch-mock` to work. I don't know why.

File: packages/snaps-simulator/src/features/simulation/slice.ts (hunk 1 above)
Comment (Mrtenz) [reply packages/snaps-simulator/src/features/simulation/slice.ts:75]: Yes, similar to the `PermissionController`, there's some weirdness with writable drafts.

File: packages/snaps-simulator/package.json (hunk 5 above)
Comment (Mrtenz) [reply packages/snaps-simulator/package.json:113]: Not sure what you mean. Which version should it be instead?

Comment (FrederikBolding) [reply packages/snaps-simulator/package.json:113]: Seems that we use `3.6.2` everywhere else?

File: packages/snaps-controllers/jest.config.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply packages/snaps-controllers/jest.config.js:14]: This is the default value though? 🤔

File: packages/snaps-simulator/webpack.config.ts (hunk 4 above)
Comment (FrederikBolding) [reply packages/snaps-simulator/webpack.config.ts:45]: Is the bundle size the same in this repo and in the snaps-simulator repo?

File: packages/snaps-simulator/.eslintrc.js (hunk 3 above)
Comment (FrederikBolding) [reply packages/snaps-simulator/.eslintrc.js:34]: Which files?

Comment (Mrtenz) [reply packages/snaps-simulator/.eslintrc.js:40]:
```
./packages/snaps-simulator/src/features/manifest/sagas.test.ts
  1:28  error  "redux-saga-test-plan" is not published  node/no-unpublished-import

./packages/snaps-simulator/src/features/polling/sagas.test.ts
  1:23  error  "jest-fetch-mock" is not published       node/no-unpublished-import
  2:28  error  "redux-saga-test-plan" is not published  node/no-unpublished-import

./packages/snaps-simulator/src/features/simulation/hooks.test.ts
  3:28  error  "redux-saga-test-plan" is not published  node/no-unpublished-import

./packages/snaps-simulator/src/features/simulation/sagas.test.ts
  4:28  error  "redux-saga-test-plan" is not published  node/no-unpublished-import

./packages/snaps-simulator/webpack.config.ts
   1:32  error  "@pmmmwh/react-refresh-webpack-plugin" is not published  node/no-unpublished-import
   2:35  error  "favicons-webpack-plugin" is not published               node/no-unpublished-import
   3:31  error  "html-webpack-plugin" is not published                   node/no-unpublished-import
   4:39  error  "monaco-editor-webpack-plugin" is not published          node/no-unpublished-import
```

File: packages/snaps-simulator/webpack.config.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      template: './src/index.html',
+    }),
+    new WebpackBarPlugin(),
+    new MonacoEditorWebpackPlugin({
+      languages: ['json', 'typescript'],
+      features: ['bracketMatching', 'clipboard', 'hover', 'unicodeHighlighter'],
+    }),
+    new FaviconsWebpackPlugin('./src/assets/favicon.svg'),
+  ],
+  // cache: {
```
Comment (FrederikBolding) [inline packages/snaps-simulator/webpack.config.ts:99]: ?

**Review by FrederikBolding — APPROVED** [review]

**Comment by Mrtenz** [comment]
@SocketSecurity ignore-all

Nothing out of the ordinary here.
